File: go/sdk/interceptors/interceptor.go
// Copyright 2025 The MCP Interceptors Authors. All rights reserved.
// Use of this source code is governed by an Apache-2.0
// license that can be found in the LICENSE file.

package interceptors

import (
	"context"
	"encoding/json"
)

// Interceptor is the common interface for all interceptors. It is implemented by both Validator and Mutator.
type Interceptor interface {
	GetMetadata() *Metadata
	GetType() InterceptorType
}

// --- Enums ---

// InterceptionPhase determines when an interceptor runs.
type InterceptionPhase string

const (
	PhaseRequest  InterceptionPhase = "request"
	PhaseResponse InterceptionPhase = "response"
	PhaseBoth     InterceptionPhase = "both"
)

// InterceptionEvent identifies a lifecycle event that can be intercepted.
type InterceptionEvent = string

// Mode controls enforcement behavior.
type Mode string

const (
	ModeEnforce Mode = "enforce" // Enforced: validation failures block, mutations apply
	ModeAudit   Mode = "audit"   // Audit: log results but don't block or apply mutations
)

// InterceptorType identifies the category of an interceptor.
type InterceptorType string

const (
	TypeValidation InterceptorType = "validation"
	TypeMutation   InterceptorType = "mutation"
)

// Severity represents validation message severity.
type Severity string

const (
	SeverityInfo  Severity = "info"
	SeverityWarn  Severity = "warn"
	SeverityError Severity = "error" // Only error blocks execution
)

// --- Metadata ---

// Compat represents protocol version compatibility.
type Compat struct {
	MinProtocol string `json:"minProtocol"`
	MaxProtocol string `json:"maxProtocol,omitempty"`
}

// Hook defines which lifecycle events and phase trigger an interceptor.
type Hook struct {
	Events []InterceptionEvent `json:"events"`
	Phase  InterceptionPhase   `json:"phase"`
}

// Metadata holds all common interceptor metadata.
type Metadata struct {
	Name         string          `json:"name"`
	Version      string          `json:"version,omitempty"`
	Description  string          `json:"description,omitempty"`
	Type         InterceptorType `json:"type"`
	Hook         Hook            `json:"hook"`
	PriorityHint Priority        `json:"priorityHint,omitempty"`
	Compat       *Compat         `json:"compat,omitempty"`
	ConfigSchema json.RawMessage `json:"configSchema,omitempty"`
	Mode         Mode            `json:"mode"`
	FailOpen     bool            `json:"failOpen,omitempty"`
}

// --- Validator ---

// ValidatorHandler is the function signature for validation handlers.
//
// Handlers MUST treat the Invocation and its Payload as read-only.
// Multiple validators for the same event run concurrently and share
// the same Invocation pointer, so any mutation of the Payload (or
// other Invocation fields) is a data race.
type ValidatorHandler func(ctx context.Context, inv *Invocation) (*ValidationResult, error)

// Validator is a validation interceptor.
type Validator struct {
	Metadata
	Handler ValidatorHandler
}

func (v *Validator) GetMetadata() *Metadata   { return &v.Metadata }
func (v *Validator) GetType() InterceptorType { return TypeValidation }

// --- Mutator ---

// MutatorHandler is the function signature for raw mutation handlers.
type MutatorHandler func(ctx context.Context, inv *Invocation) (*MutationResult, error)

// Mutator is a mutation interceptor.
type Mutator struct {
	Metadata
	Handler MutatorHandler
}

func (m *Mutator) GetMetadata() *Metadata   { return &m.Metadata }
func (m *Mutator) GetType() InterceptorType { return TypeMutation }

// --- Invocation ---

// Invocation is the context passed to every interceptor handler.
type Invocation struct {
	Event   string             // e.g. "tools/call"
	Phase   InterceptionPhase  // "request" or "response"
	Payload any                // The payload (json.RawMessage when invoked via interceptor/invoke)
	Config  map[string]any     // Per-invocation config
	Context *InvocationContext // Optional caller context (identity, trace, etc.)
}

// InvocationContext holds optional context passed to interceptors.
type InvocationContext struct {
	Principal *Principal `json:"principal,omitempty"`
	TraceID   string     `json:"traceId,omitempty"`
	SpanID    string     `json:"spanId,omitempty"`
	Timestamp string     `json:"timestamp,omitempty"`
	SessionID string     `json:"sessionId,omitempty"`
}

// Principal identifies the caller.
type Principal struct {
	Type   string         `json:"type"`
	ID     string         `json:"id,omitempty"`
	Claims map[string]any `json:"claims,omitempty"`
}

// --- Results ---

// ValidationMessage is a single validation finding.
type ValidationMessage struct {
	Path     string   `json:"path,omitempty"`
	Message  string   `json:"message"`
	Severity Severity `json:"severity"`
}

// ValidationSuggestion is an optional suggested correction.
type ValidationSuggestion struct {
	Path  string `json:"path"`
	Value any    `json:"value"`
}

// ValidationResult is returned by validation interceptors.
type ValidationResult struct {
	Valid       bool                   `json:"valid"`
	Severity    Severity               `json:"severity,omitempty"`
	Messages    []ValidationMessage    `json:"messages,omitempty"`
	Suggestions []ValidationSuggestion `json:"suggestions,omitempty"`
}

// MutationResult is returned by mutation interceptors.
type MutationResult struct {
	Modified bool            `json:"modified"`
	Info     map[string]any  `json:"info,omitempty"`
	Payload  json.RawMessage `json:"payload,omitempty"`
}

File: python/sdk/src/mcp_ext_interceptors/interceptor.py
# Copyright 2025 The MCP Interceptors Authors. All rights reserved.
# Use of this source code is governed by a Apache-2.0
# license that can be found in the LICENSE file.

"""Author-side API for writing interceptors (SEP-2624).

Handlers receive an :class:`Invocation` and return the type-specific result
fields; the hosting side stamps the common wire fields (`interceptor`,
`phase`, `durationMs`) before returning them to the invoker.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass
from typing import Any

from mcp_ext_interceptors.types import (
    TYPE_MUTATION,
    TYPE_VALIDATION,
    Hook,
    InterceptorInfo,
    InvocationContext,
    MutationResult,
    Phase,
    ValidationResult,
)


@dataclass(kw_only=True, frozen=True)
class Invocation:
    """Read-only view of one interceptor invocation, passed to handlers.

    Treat `payload` as immutable: validators for the same event run
    concurrently and share it. Mutators return a new payload in their result
    instead of modifying this one in place.
    """

    event: str
    phase: Phase
    payload: Any
    config: dict[str, Any] | None = None
    context: InvocationContext | None = None


ValidatorHandler = Callable[[Invocation], Awaitable[ValidationResult]]
MutatorHandler = Callable[[Invocation], Awaitable[MutationResult]]


@dataclass(kw_only=True)
class Validator:
    """A validation interceptor: descriptor plus its handler."""

    info: InterceptorInfo
    handler: ValidatorHandler

    def __post_init__(self) -> None:
        if self.info.type != TYPE_VALIDATION:
            raise ValueError(f"Validator {self.info.name!r} must have type {TYPE_VALIDATION!r}, got {self.info.type!r}")


@dataclass(kw_only=True)
class Mutator:
    """A mutation interceptor: descriptor plus its handler."""

    info: InterceptorInfo
    handler: MutatorHandler

    def __post_init__(self) -> None:
        if self.info.type != TYPE_MUTATION:
            raise ValueError(f"Mutator {self.info.name!r} must have type {TYPE_MUTATION!r}, got {self.info.type!r}")


RegisteredInterceptor = Validator | Mutator


def build_hooks(
    *,
    events: Sequence[str] | None,
    phase: Phase | None,
    hooks: Sequence[Hook] | None,
) -> list[Hook]:
    """Build the hooks list from either `events`+`phase` or explicit `hooks`."""
    if hooks is not None:
        if events is not None or phase is not None:
            raise ValueError("pass either hooks= or events=/phase=, not both")
        if not hooks:
            raise ValueError("hooks must not be empty")
        return list(hooks)
    if events is None or phase is None:
        raise ValueError("events= and phase= are required when hooks= is not given")
    if not events:
        raise ValueError("events must not be empty")
    return [Hook(events=list(events), phase=phase)]
